File: apps/web/src/shared/validation/multimodal.ts
import { matchesApiSchema } from '@/shared/api/generated/assistant';
import type {
  AnalyticalMapFeature,
  CommonOperationalPicture,
  CopGeometry,
  CopLayer,
  MultimodalEvidenceState,
  SourceMapFeature,
} from '@/shared/types/assistant';

export function isMultimodalEvidenceState(
  value: unknown,
): value is MultimodalEvidenceState {
  if (!matchesApiSchema('MultimodalStateResponse', value)) return false;
  const state = value as MultimodalEvidenceState;
  return (
    Array.isArray(state.assets) &&
    Array.isArray(state.associations) &&
    Array.isArray(state.observations) &&
    state.assets.every(
      (asset) =>
        nonEmpty(
          asset.asset_id,
          asset.retrieved_at,
          asset.modality,
          asset.media_type,
          asset.content_sha256,
          asset.capture_role,
          asset.eligibility,
          asset.source.source_id,
          asset.source.attribution,
        ) && asset.byte_length > 0,
    ) &&
    state.associations.every((association) =>
      nonEmpty(
        association.association_id,
        association.asset_id,
        association.physical_event_id,
        association.status,
        association.detail,
      ),
    ) &&
    state.observations.every(
      (observation) =>
        observation.modality === 'image' &&
        observation.truth_status === 'analytical' &&
        nonEmpty(
          observation.observation_id,
          observation.asset_id,
          observation.association_id,
          observation.physical_event_id,
          observation.kind,
          observation.status,
          observation.uncertainty,
          observation.created_at,
          observation.configuration.model_id,
          observation.configuration.adapter_version,
          observation.configuration.analysis_version,
          observation.configuration.prompt_version,
          observation.configuration.preprocessing_version,
        ) &&
        observation.configuration.maximum_output_tokens > 0,
    ) &&
    referencesStayWithinState(state)
  );
}

export function isCommonOperationalPicture(
  value: unknown,
): value is CommonOperationalPicture {
  if (!matchesApiSchema('CommonOperationalPictureResponse', value)) return false;
  const cop = value as CommonOperationalPicture;
  return (
    cop.layers.length > 0 &&
    cop.layers.every(
      (layer) =>
        layer.physical_event_id === cop.physical_event_id && layerIsConsistent(layer),
    )
  );
}

export function copMatchesMultimodalState(
  cop: CommonOperationalPicture,
  state: MultimodalEvidenceState,
): boolean {
  if (
    cop.multimodal_state_version !== state.state_version ||
    cop.physical_event_id !== state.physical_event_id
  ) {
    return false;
  }
  const assets = new Set(state.assets.map((item) => item.asset_id));
  const observations = new Set(state.observations.map((item) => item.observation_id));
  return cop.layers.every((layer) =>
    layer.features.every(
      (feature) =>
        feature.source_asset_ids.every((id) => assets.has(id)) &&
        (feature.feature_type === 'source' ||
          feature.visual_observation_ids.every((id) => observations.has(id))),
    ),
  );
}

function referencesStayWithinState(state: MultimodalEvidenceState): boolean {
  const assets = new Set(state.assets.map((item) => item.asset_id));
  const associations = new Set(state.associations.map((item) => item.association_id));
  return (
    state.associations.every(
      (item) =>
        assets.has(item.asset_id) && item.physical_event_id === state.physical_event_id,
    ) &&
    state.observations.every(
      (item) =>
        assets.has(item.asset_id) &&
        associations.has(item.association_id) &&
        item.physical_event_id === state.physical_event_id,
    )
  );
}

function layerIsConsistent(layer: CopLayer): boolean {
  if (
    layer.features.length === 0 ||
    !nonEmpty(
      layer.layer_id,
      layer.physical_event_id,
      layer.title,
      layer.semantic_kind,
      layer.created_at,
      layer.updated_at,
      layer.status,
      layer.uncertainty,
      layer.attribution,
    )
  ) {
    return false;
  }
  if (layer.layer_type === 'source') {
    return (
      layer.source_ids.length > 0 &&
      layer.source_asset_ids.length > 0 &&
      layer.features.every(
        (feature) =>
          sourceFeatureIsConsistent(feature) &&
          feature.physical_event_id === layer.physical_event_id,
      ) &&
      sameStrings(
        layer.source_ids,
        layer.features.map((feature) => feature.source_id),
      ) &&
      sameStrings(
        layer.source_asset_ids,
        layer.features.flatMap((feature) => feature.source_asset_ids),
      )
    );
  }
  return (
    layer.source_asset_ids.length > 0 &&
    layer.visual_observation_ids.length > 0 &&
    layer.features.every(
      (feature) =>
        analyticalFeatureIsConsistent(feature) &&
        feature.physical_event_id === layer.physical_event_id,
    ) &&
    sameStrings(
      layer.source_asset_ids,
      layer.features.flatMap((feature) => feature.source_asset_ids),
    ) &&
    sameStrings(
      layer.visual_observation_ids,
      layer.features.flatMap((feature) => feature.visual_observation_ids),
    )
  );
}

function sourceFeatureIsConsistent(feature: SourceMapFeature): boolean {
  const authorityMatches =
    (feature.source_authority === 'official' &&
      feature.authority === 'official_source') ||
    (feature.source_authority === 'source_supplied' &&
      feature.authority === 'source_supplied');
  return (
    authorityMatches && nonEmpty(feature.source_id) && featureBaseIsConsistent(feature)
  );
}

function analyticalFeatureIsConsistent(feature: AnalyticalMapFeature): boolean {
  return (
    feature.authority === 'analytical_generated' &&
    feature.visual_observation_ids.length > 0 &&
    featureBaseIsConsistent(feature)
  );
}

function featureBaseIsConsistent(
  feature: SourceMapFeature | AnalyticalMapFeature,
): boolean {
  return (
    nonEmpty(
      feature.feature_id,
      feature.physical_event_id,
      feature.created_at,
      feature.semantic_kind,
      feature.attribution,
      feature.status,
      feature.uncertainty,
    ) &&
    feature.source_asset_ids.length > 0 &&
    geometryIsValid(feature.geometry)
  );
}

function geometryIsValid(value: CopGeometry): boolean {
  if (value.crs !== 'EPSG:4326') return false;
  if (value.type === 'Point') return isCoordinate(value.coordinates);
  if (value.type === 'LineString') {
    return (
      value.coordinates.length >= 2 &&
      value.coordinates.length <= 4_096 &&
      value.coordinates.every(isCoordinate)
    );
  }
  return isPolygonCoordinates(value.coordinates);
}

function isCoordinate(value: [number, number]): boolean {
  return (
    value.every(Number.isFinite) &&
    value[0] >= -180 &&
    value[0] <= 180 &&
    value[1] >= -90 &&
    value[1] <= 90
  );
}

function isPolygonCoordinates(value: [number, number][][]): boolean {
  if (value.length === 0 || value.length > 8) return false;
  if (
    value.some(
      (ring) =>
        ring.length < 4 ||
        !ring.every(isCoordinate) ||
        ring[0][0] !== ring[ring.length - 1][0] ||
        ring[0][1] !== ring[ring.length - 1][1],
    )
  ) {
    return false;
  }
  if (
    value.reduce((total, ring) => total + ring.length, 0) > 4_096 ||
    value.some((ring) => Math.abs(signedArea(ring)) < 1e-12 || selfIntersects(ring))
  ) {
    return false;
  }
  const [exterior, ...holes] = value;
  if (
    holes.some(
      (hole) =>
        ringsIntersect(exterior, hole) ||
        hole.slice(0, -1).some((point) => !pointInRing(point, exterior)),
    )
  ) {
    return false;
  }
  return !holes.some((first, index) =>
    holes
      .slice(index + 1)
      .some(
        (second) =>
          ringsIntersect(first, second) ||
          pointInRing(first[0], second) ||
          pointInRing(second[0], first),
      ),
  );
}

function signedArea(ring: [number, number][]): number {
  return (
    ring
      .slice(0, -1)
      .reduce(
        (area, point, index) =>
          area + point[0] * ring[index + 1][1] - ring[index + 1][0] * point[1],
        0,
      ) / 2
  );
}

function orientation(
  first: [number, number],
  second: [number, number],
  third: [number, number],
): number {
  return (
    (second[0] - first[0]) * (third[1] - first[1]) -
    (second[1] - first[1]) * (third[0] - first[0])
  );
}

function onSegment(
  first: [number, number],
  second: [number, number],
  point: [number, number],
): boolean {
  return (
    point[0] >= Math.min(first[0], second[0]) &&
    point[0] <= Math.max(first[0], second[0]) &&
    point[1] >= Math.min(first[1], second[1]) &&
    point[1] <= Math.max(first[1], second[1]) &&
    Math.abs(orientation(first, second, point)) < 1e-12
  );
}

function segmentsIntersect(
  first: [number, number],
  second: [number, number],
  third: [number, number],
  fourth: [number, number],
): boolean {
  const values = [
    orientation(first, second, third),
    orientation(first, second, fourth),
    orientation(third, fourth, first),
    orientation(third, fourth, second),
  ];
  if (values[0] * values[1] < 0 && values[2] * values[3] < 0) return true;
  return (
    (Math.abs(values[0]) < 1e-12 && onSegment(first, second, third)) ||
    (Math.abs(values[1]) < 1e-12 && onSegment(first, second, fourth)) ||
    (Math.abs(values[2]) < 1e-12 && onSegment(third, fourth, first)) ||
    (Math.abs(values[3]) < 1e-12 && onSegment(third, fourth, second))
  );
}

function selfIntersects(ring: [number, number][]): boolean {
  const lastSegment = ring.length - 2;
  for (let first = 0; first < ring.length - 1; first += 1) {
    for (let second = first + 1; second < ring.length - 1; second += 1) {
      if (second === first + 1 || (first === 0 && second === lastSegment)) continue;
      if (
        segmentsIntersect(ring[first], ring[first + 1], ring[second], ring[second + 1])
      ) {
        return true;
      }
    }
  }
  return false;
}

function ringsIntersect(
  first: [number, number][],
  second: [number, number][],
): boolean {
  return first
    .slice(0, -1)
    .some((point, firstIndex) =>
      second
        .slice(0, -1)
        .some((other, secondIndex) =>
          segmentsIntersect(
            point,
            first[firstIndex + 1],
            other,
            second[secondIndex + 1],
          ),
        ),
    );
}

function pointInRing(point: [number, number], ring: [number, number][]): boolean {
  if (
    ring.slice(0, -1).some((first, index) => onSegment(first, ring[index + 1], point))
  ) {
    return false;
  }
  let inside = false;
  for (let index = 0; index < ring.length - 1; index += 1) {
    const first = ring[index];
    const second = ring[index + 1];
    if (first[1] > point[1] === second[1] > point[1]) continue;
    const boundary =
      first[0] +
      ((point[1] - first[1]) * (second[0] - first[0])) / (second[1] - first[1]);
    if (point[0] < boundary) inside = !inside;
  }
  return inside;
}

function nonEmpty(...values: string[]): boolean {
  return values.every((value) => value.trim().length > 0);
}

function sameStrings(left: string[], right: string[]): boolean {
  return (
    new Set(left).size === new Set(right).size &&
    [...new Set(left)].every((item) => new Set(right).has(item))
  );
}

File: apps/api/src/disaster_monitor/domain/multimodal.py
"""Typed multimodal evidence and common-operational-picture artifacts."""

from dataclasses import dataclass, field
from datetime import datetime
from enum import StrEnum
from math import isfinite


class AssetModality(StrEnum):
    """Raw media classes accepted by the bounded multimodal boundary."""

    IMAGE = "image"


class CaptureRole(StrEnum):
    """Temporal role asserted by source metadata, never inferred from pixels."""

    PRE_EVENT = "pre_event"
    POST_EVENT = "post_event"
    SINGLE_CAPTURE = "single_capture"
    UNKNOWN = "unknown"


class AssetEligibility(StrEnum):
    """Whether an asset has enough provenance for grounded visual analysis."""

    ANALYSIS_ELIGIBLE = "analysis_eligible"
    PREVIEW_ONLY = "preview_only"
    ORPHANED = "orphaned"
    REJECTED = "rejected"


class EventAssociationStatus(StrEnum):
    """Deterministic relationship between an asset and a physical event."""

    ASSOCIATED = "associated"
    AMBIGUOUS = "ambiguous"
    UNMATCHED = "unmatched"
    ORPHANED = "orphaned"


class VisualObservationKind(StrEnum):
    """Canonical analytical products emitted by visual analysis."""

    DAMAGE_ASSESSMENT = "damage_assessment"
    VISUAL_QUESTION_ANSWER = "visual_question_answer"
    OCR_EXTRACTION = "ocr_extraction"


class DamageLevel(StrEnum):
    """Roadmap-aligned ordinal visible-damage categories."""

    NO_VISIBLE_DAMAGE = "no_visible_damage"
    MINOR_DAMAGE = "minor_damage"
    MAJOR_DAMAGE = "major_damage"
    DESTROYED = "destroyed"
    UNKNOWN = "unknown"


class VisualObservationStatus(StrEnum):
    """Outcome of one bounded analytical request."""

    PRODUCED = "produced"
    ABSTAINED = "abstained"
    RESEARCH_ONLY = "research_only"


class VisualTruthStatus(StrEnum):
    """Epistemic marker that cannot be confused with a reported fact."""

    ANALYTICAL = "analytical"


class GeometryKind(StrEnum):
    POINT = "point"
    LINE_STRING = "line_string"
    POLYGON = "polygon"


class SourceGeometryAuthority(StrEnum):
    """Authority that may only be used by source-supplied map features."""

    OFFICIAL = "official"
    SOURCE_SUPPLIED = "source_supplied"


class MapFeatureAuthority(StrEnum):
    """Machine-readable distinction between source and generated geometry."""

    OFFICIAL_SOURCE = "official_source"
    SOURCE_SUPPLIED = "source_supplied"
    ANALYTICAL_GENERATED = "analytical_generated"


class CopStatus(StrEnum):
    CURRENT = "current"
    STALE = "stale"
    PREVIEW_ONLY = "preview_only"


@dataclass(frozen=True, slots=True)
class GeoPoint:
    """One WGS84 longitude/latitude coordinate."""

    longitude: float
    latitude: float

    def __post_init__(self) -> None:
        if not isfinite(self.longitude) or not isfinite(self.latitude):
            raise ValueError("Coordinates must be finite.")
        if not -180 <= self.longitude <= 180:
            raise ValueError("Longitude must be between -180 and 180.")
        if not -90 <= self.latitude <= 90:
            raise ValueError("Latitude must be between -90 and 90.")


@dataclass(frozen=True, slots=True)
class GeoLineString:
    """Validated coordinate sequence in WGS84."""

    points: tuple[GeoPoint, ...]
    crs: str = "EPSG:4326"
    kind: GeometryKind = field(init=False, default=GeometryKind.LINE_STRING)

    def __post_init__(self) -> None:
        if self.crs != "EPSG:4326":
            raise ValueError("Only explicit EPSG:4326 geometry is supported.")
        if len(self.points) < 2:
            raise ValueError("A line string requires at least two points.")


@dataclass(frozen=True, slots=True)
class GeoPolygon:
    """One WGS84 polygon with a closed exterior and optional closed holes."""

    rings: tuple[tuple[GeoPoint, ...], ...]
    crs: str = "EPSG:4326"
    kind: GeometryKind = field(init=False, default=GeometryKind.POLYGON)

    def __post_init__(self) -> None:
        if self.crs != "EPSG:4326":
            raise ValueError("Only explicit EPSG:4326 geometry is supported.")
        if not self.rings:
            raise ValueError("A polygon requires an exterior ring.")
        for ring in self.rings:
            if len(ring) < 4 or ring[0] != ring[-1]:
                raise ValueError(
                    "Every polygon ring requires at least four closed coordinates."
                )


MapGeometry = GeoPoint | GeoLineString | GeoPolygon


@dataclass(frozen=True, slots=True)
class MultimodalSourceMetadata:
    """Audit metadata for an admitted source asset."""

    source_id: str
    attribution: str
    canonical_url: str | None = None
    dataset_id: str | None = None
    license_name: str | None = None

    def __post_init__(self) -> None:
        if not self.source_id.strip():
            raise ValueError("Multimodal source metadata requires a stable source ID.")
        if not self.attribution.strip():
            raise ValueError("Multimodal source metadata requires attribution text.")
        if self.canonical_url is not None and not self.canonical_url.startswith(
            "https://"
        ):
            raise ValueError("A supplied canonical asset URL must use HTTPS.")


@dataclass(frozen=True, slots=True)
class MultimodalAsset:
    """A raw source asset, kept separate from all AI interpretations."""

    asset_id: str
    source: MultimodalSourceMetadata
    retrieved_at: datetime
    captured_at: datetime | None
    modality: AssetModality
    media_type: str
    content_sha256: str
    byte_length: int
    width: int | None
    height: int | None
    footprint: GeoPolygon | None
    declared_hazard: StrEnum | None
    declared_country_code: str | None
    capture_role: CaptureRole
    processing_level: str | None
    parent_asset_ids: tuple[str, ...]
    event_id_hint: str | None
    eligibility: AssetEligibility
    eligibility_reasons: tuple[str, ...]
    content: bytes = field(repr=False, compare=False)

    def __post_init__(self) -> None:
        if not self.asset_id.strip():
            raise ValueError("A multimodal asset requires a stable asset ID.")
        if len(self.content_sha256) != 64 or any(
            character not in "0123456789abcdef" for character in self.content_sha256
        ):
            raise ValueError("A multimodal asset requires a lowercase SHA-256 hash.")
        if self.byte_length <= 0 or self.byte_length != len(self.content):
            raise ValueError("Asset byte length must match non-empty content.")
        if (self.width is None) != (self.height is None):
            raise ValueError("Image width and height must both be known or unknown.")
        if self.width is not None and (self.width <= 0 or (self.height or 0) <= 0):
            raise ValueError("Image dimensions must be positive.")


@dataclass(frozen=True, slots=True)
class AssetEventAssociation:
    """Auditable, metadata-owned association to one physical event."""

    association_id: str
    asset_id: str
    physical_event_id: str
    status: EventAssociationStatus
    geography_match: bool | None
    time_match: bool | None
    hazard_match: bool | None
    country_match: bool | None
    event_id_match: bool | None
    distance_km: float | None
    time_delta_seconds: float | None
    rule_ids: tuple[str, ...]
    detail: str


@dataclass(frozen=True, slots=True)
class VisualAnalysisConfiguration:
    """Reproducible model, prompt, preprocessing, and decoding identity."""

    model_id: str
    model_digest: str | None
    adapter_version: str
    analysis_version: str
    prompt_version: str
    preprocessing_version: str
    maximum_output_tokens: int
    temperature: float
    seed: int

    def __post_init__(self) -> None:
        if self.maximum_output_tokens <= 0:
            raise ValueError("Visual analysis requires a positive output-token cap.")


@dataclass(frozen=True, slots=True)
class VisualObservation:
    """An AI-derived visual assertion that can never become a ReportedFact."""

    observation_id: str
    asset_id: str
    association_id: str
    physical_event_id: str
    kind: VisualObservationKind
    status: VisualObservationStatus
    damage_level: DamageLevel | None
    question: str | None
    answer: str | None
    answerable: bool | None
    confidence: float | None
    uncertainty: str
    visual_cues: tuple[str, ...]
    configuration: VisualAnalysisConfiguration
    created_at: datetime
    safety_rule_ids: tuple[str, ...] = ()
    truth_status: VisualTruthStatus = VisualTruthStatus.ANALYTICAL

    def __post_init__(self) -> None:
        if not self.asset_id or not self.association_id or not self.physical_event_id:
            raise ValueError("Visual observations require complete artifact lineage.")
        if not self.uncertainty.strip():
            raise ValueError("Visual observations require explicit uncertainty text.")
        if self.confidence is not None and not 0 <= self.confidence <= 1:
            raise ValueError("Visual confidence must be between zero and one.")
        if self.status == VisualObservationStatus.ABSTAINED and self.answerable is True:
            raise ValueError("An abstained visual answer cannot be marked answerable.")
        if (
            self.kind == VisualObservationKind.DAMAGE_ASSESSMENT
            and self.status == VisualObservationStatus.ABSTAINED
            and self.damage_level != DamageLevel.UNKNOWN
        ):
            raise ValueError("An abstained damage assessment must remain unknown.")


@dataclass(frozen=True, slots=True)
class MultimodalEvidenceState:
    """Versioned analytical extension of, not replacement for, EW state."""

    state_version: str
    evidence_world_state_version: str
    physical_event_id: str
    assets: tuple[MultimodalAsset, ...]
    associations: tuple[AssetEventAssociation, ...]
    observations: tuple[VisualObservation, ...]
    evaluated_at: datetime


@dataclass(frozen=True, slots=True)
class SourceMapFeature:
    """Geometry supplied by a named source; never generated by the model."""

    feature_id: str
    physical_event_id: str
    source_id: str
    source_asset_ids: tuple[str, ...]
    created_at: datetime
    updated_at: datetime | None
    semantic_kind: str
    geometry: MapGeometry
    attribution: str
    status: CopStatus
    uncertainty: str
    source_authority: SourceGeometryAuthority
    authority: MapFeatureAuthority = field(init=False)

    def __post_init__(self) -> None:
        authority = (
            MapFeatureAuthority.OFFICIAL_SOURCE
            if self.source_authority == SourceGeometryAuthority.OFFICIAL
            else MapFeatureAuthority.SOURCE_SUPPLIED
        )
        object.__setattr__(self, "authority", authority)
        if not self.source_id.strip() or not self.source_asset_ids:
            raise ValueError("Source map features require source and asset provenance.")
        if not self.attribution.strip() or not self.uncertainty.strip():
            raise ValueError(
                "Source map features require visible attribution and uncertainty."
            )


@dataclass(frozen=True, slots=True)
class AnalyticalMapFeature:
    """Generated analytical geometry that is structurally never official."""

    feature_id: str
    physical_event_id: str
    source_asset_ids: tuple[str, ...]
    visual_observation_ids: tuple[str, ...]
    created_at: datetime
    updated_at: datetime | None
    semantic_kind: str
    geometry: MapGeometry
    attribution: str
    status: CopStatus
    uncertainty: str
    confidence: float | None
    authority: MapFeatureAuthority = field(
        init=False, default=MapFeatureAuthority.ANALYTICAL_GENERATED
    )

    def __post_init__(self) -> None:
        if not self.source_asset_ids or not self.visual_observation_ids:
            raise ValueError(
                "Analytical map features require asset and observation provenance."
            )
        if not self.attribution.strip() or not self.uncertainty.strip():
            raise ValueError(
                "Analytical map features require visible attribution and uncertainty."
            )
        if self.confidence is not None and not 0 <= self.confidence <= 1:
            raise ValueError(
                "Analytical feature confidence must be between zero and one."
            )


@dataclass(frozen=True, slots=True)
class SourceMapLayer:
    """Layer whose features all originate from explicit source geometry."""

    layer_id: str
    physical_event_id: str
    title: str
    semantic_kind: str
    features: tuple[SourceMapFeature, ...]
    source_ids: tuple[str, ...]
    source_asset_ids: tuple[str, ...]
    created_at: datetime
    updated_at: datetime
    status: CopStatus
    uncertainty: str
    attribution: str

    def __post_init__(self) -> None:
        if not self.features or any(
            feature.physical_event_id != self.physical_event_id
            for feature in self.features
        ):
            raise ValueError("Source layers require same-event source features.")
        if set(self.source_ids) != {item.source_id for item in self.features}:
            raise ValueError("Source layer source lineage must match its features.")
        if set(self.source_asset_ids) != {
            asset_id for item in self.features for asset_id in item.source_asset_ids
        }:
            raise ValueError("Source layer asset lineage must match its features.")
        if not self.semantic_kind.strip() or not self.uncertainty.strip():
            raise ValueError("Source layers require semantics and uncertainty.")


@dataclass(frozen=True, slots=True)
class AnalyticalMapLayer:
    """Layer whose features are explicitly generated analytical products."""

    layer_id: str
    physical_event_id: str
    title: str
    semantic_kind: str
    features: tuple[AnalyticalMapFeature, ...]
    source_asset_ids: tuple[str, ...]
    visual_observation_ids: tuple[str, ...]
    created_at: datetime
    updated_at: datetime
    status: CopStatus
    uncertainty: str
    attribution: str

    def __post_init__(self) -> None:
        if not self.features or any(
            feature.physical_event_id != self.physical_event_id
            for feature in self.features
        ):
            raise ValueError(
                "Analytical layers require same-event analytical features."
            )
        if set(self.source_asset_ids) != {
            asset_id for item in self.features for asset_id in item.source_asset_ids
        }:
            raise ValueError("Analytical layer asset lineage must match its features.")
        if set(self.visual_observation_ids) != {
            observation_id
            for item in self.features
            for observation_id in item.visual_observation_ids
        }:
            raise ValueError(
                "Analytical layer observation lineage must match its features."
            )
        if not self.semantic_kind.strip() or not self.uncertainty.strip():
            raise ValueError("Analytical layers require semantics and uncertainty.")


CopLayer = SourceMapLayer | AnalyticalMapLayer


@dataclass(frozen=True, slots=True)
class CommonOperationalPicture:
    """Renderer-independent, provenance-bearing map product."""

    cop_id: str
    physical_event_id: str
    multimodal_state_version: str
    created_at: datetime
    updated_at: datetime
    status: CopStatus
    layers: tuple[CopLayer, ...]

    def __post_init__(self) -> None:
        if any(
            layer.physical_event_id != self.physical_event_id for layer in self.layers
        ):
            raise ValueError("Every COP layer must describe the same physical event.")
